cc/pass_alive_test.cc

```diff
@@ -0,0 +1,732 @@
+      Coord c(row, col);
+      auto color = stones[c];
+      if (color == Color::kWhite) {
+        oss << print_white << "O ";
+      } else if (color == Color::kBlack) {
+        oss << print_black << "X ";
+      } else {
+        oss << print_empty << ". ";
+      }
+    }
+    oss << print_normal << absl::StreamFormat("%2d", kN - row);
+    oss << "\n";
+  }
+  format_cols();
+  return oss.str();
+}
+
+class PassAliveTest : public ::testing::Test {
+ protected:
+  struct TestCase {
+    TestCase(const std::string& board, const std::string& expected)
+        : board(board), expected(ParseBoard(expected)) {}
+
+    TestablePosition board;
+    std::array<Color, kN * kN> expected;
+  };
+
+  void RunTests(absl::Span<const TestCase> tests) {
+    for (const auto& test : tests) {
+      MG_LOG(INFO) << "board state:\n" << test.board.ToPrettyString();
+      MG_LOG(INFO) << "Calculating black pass-alive regions";
+      auto black = test.board.CalculatePassAliveRegions(Color::kBlack);
+
+      MG_LOG(INFO) << "Calculating white pass-alive regions";
+      auto white = test.board.CalculatePassAliveRegions(Color::kWhite);
+
+      // Initialize the result to the input board state.
+      std::array<Color, kN * kN> actual;
+      for (size_t i = 0; i < kN * kN; ++i) {
+        actual[i] = test.board.stones()[i].color();
+      }
+
+      // Merge both pass-alive regions into the result.
+      for (size_t i = 0; i < kN * kN; ++i) {
+        MG_CHECK(black[i] == Color::kEmpty || white[i] == Color::kEmpty)
+            << Coord(i).ToGtp()
+            << " was marked as belonging to both black & white pass-alive "
+               "regions";
+        if (black[i] != Color::kEmpty) {
+          actual[i] = Color::kBlack;
+        } else if (white[i] != Color::kEmpty) {
+          actual[i] = Color::kWhite;
+        }
+      }
+
+      MG_CHECK(test.expected == actual)
+          << "\nexpected:\n"
+          << ToPrettyString(test.expected) << "\n\nactual:\n"
+          << ToPrettyString(actual) << "\n";
+
+      MG_LOG(INFO) << "expected:\n" << ToPrettyString(test.expected);
+      MG_LOG(INFO) << "actual:\n" << ToPrettyString(actual);
+    }
+  }
+};
+
+TEST_F(PassAliveTest, 9x9) {
+  if (kN != 9) {
+    return;
+  }
+
+  TestCase tests[] = {
+      {// board state
+       R"(.X.XO....
+          XXXXO....
+          OOOOO....
+          .........
+          .........
+          .........
+          .........
+          .........
+          .........)",
+       // expected result
+       R"(XXXXO....
+          XXXXO....
+          OOOOO....
+          .........
+          .........
+          .........
+          .........
+          .........
+          .........)"},
+
+      {// board state
+       R"(....O.O..
+          ....OOO..
+          ......OOO
+          ......O..
+          ......OOO
+          ......OXX
+          ......OX.
+          ......OXX
+          ......OX.)",
+       // expected result
+       R"(....OOOOO
```

Comment:
Why is the corner pass alive?

If we make it 2x5 it should not be pass alive

Reply:
not sure, but i think this vs. the example on line 214 is relevant

Reply:
Yeah, the two tests on lines 194 & 214 give more examples of this behaviour.